namecoin/src/main/java/org/libdohj/names/NameLookupLatestLevelDBTransactionCache.java

```diff
@@ -0,0 +1,410 @@
+        
+        try {
+            tryOpen(directory, dbFactory, options);
+        } catch (IOException e) {
+            try {
+                dbFactory.repair(directory, options);
+                tryOpen(directory, dbFactory, options);
+            } catch (IOException e1) {
+                throw new Exception(e1);
+            }
+        }
+        
+        chain.addNewBestBlockListener(Threading.SAME_THREAD, this);
+        chain.addReorganizeListener(Threading.SAME_THREAD, this);
+        chain.addTransactionReceivedListener(Threading.SAME_THREAD, this);
+    }
+    
+    protected void tryOpen(File directory, DBFactory dbFactory, Options options) throws IOException {
+        db = dbFactory.open(directory, options);
+        initStoreIfNeeded();
+    }
+    
+    protected synchronized void initStoreIfNeeded() {
+        if (db.get(CHAIN_HEAD_KEY) != null)
+            return;   // Already initialised.
+        
+        setChainHead(0);
+    }
+    
+    protected StoredBlock getSafeBlock(StoredBlock block) throws BlockStoreException {
+        
+        StoredBlock result = block;
+        
+        int safetyCount;
+        for (safetyCount = 0; safetyCount < 12; safetyCount++) {
+            result = result.getPrev(store);
+        }
+        
+        return result;
+    }
+    
+    protected synchronized void putBlockChain(StoredBlock block) throws Exception {
+        
+        // TODO: use BIP 113 timestamps
+        if ( (new Date().getTime() / 1000 ) - block.getHeader().getTimeSeconds() > 366 * 24 * 60 * 60) {
+            log.debug("NameDB halting walkbalk due to timestamp expiration, height " + block.getHeight());
+            return;
+        }
+        
+        if (block.getHeight() > getChainHead() + 1) {
+            putBlockChain(block.getPrev(store));
+        }
+        
+        putBlock(block);
+    }
+    
+    // TODO: try a different peer if downloading a block fails, otherwise we're likely to stall the syncup
+    protected synchronized void putBlock(StoredBlock block) throws Exception {
+        
+        Sha256Hash blockHash = block.getHeader().getHash();
+        
+        // We might not have the block's transactions already; if we don't, we have to download the block again.
+        // This should be very rare; I'm not actually certain what circumstances would trigger it.
+        // (I guess it would happen if block.transactions is null?)
+        if (! pendingBlockTransactions.containsKey(block.getHeader().getHash())) {
+            log.warn("Transactions missing from block " + blockHash + "; re-downloading block...");
+            
+            Block nameFullBlock = peerGroup.getDownloadPeer().getBlock(blockHash).get();
+            
+            // The full block hasn't been verified in any way!
+            // So let's do that now.
+            
+            if (! nameFullBlock.getHash().equals(blockHash)) {
+                throw new Exception("Block hash mismatch!");
```

Comment:
Please use `VerificationException` instead of plain `Exception`, which means we then don't have to catch and rethrow further up.